smartypants.py

```diff
@@ -563,11 +569,15 @@ def _tokenize(text):
     Based on the _tokenize() subroutine from `Brad Choate's MTRegex plugin`__.
 
     __ http://www.bradchoate.com/past/mtregex.php
-    """
 
+
+    If you have the ``regex`` library (https://pypi.python.org/pypi/regex/),
+    this function will use an alternative regex to perform significantly faster
+    on large input texts.
+    """
     tokens = []
 
-    tag_soup = re.compile(r'([^<]*)(<!--.*?--\s*>|<[^>]*>)', re.S)
+    tag_soup = re.compile(TAG_REGEX, re.S)
 
     token_match = tag_soup.search(text)
```

Comment:
As far as I understand, this is the part where backtracking happens. It's not the pattern itself that goes back, it's the `.search`, which makes python retry the `[^<]*`match starting from each of the string characters:

But since our pattern starts with a wildcard match and we rerun it after matching a tag from the end of previous match I think we can replace this (and the repeated call within the loop) with `.match`, which will only try to apply the pattern from the beginning of the string. But I might be missing something about why we need `.search` there.

```
In [24]: %timeit re.search(r'([^<]*)(<!--.*?--\s*>|<[^>]*>)', 'hello'*2000)
1.02 s ± 71.1 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [25]: %timeit re.match(r'([^<]*)(<!--.*?--\s*>|<[^>]*>)', 'hello'*2000)
202 µs ± 4.92 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)

# This doesn't happen when there is a match

In [26]: %timeit re.match(r'([^<]*)(<!--.*?--\s*>|<[^>]*>)', 'hello<tag>'*2000)
2.34 µs ± 214 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

In [27]: %timeit re.search(r'([^<]*)(<!--.*?--\s*>|<[^>]*>)', 'hello<tag>'*2000)
2.17 µs ± 113 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
```